File: algorithms/sorting/bubble-sort/implementations/javascript/bubbleSort.js
/**
 * Bubble Sort Implementation in JavaScript
 * 冒泡排序的JavaScript实现
 * 
 * Time Complexity: O(n²)
 * Space Complexity: O(1)
 */

/**
 * Sorts an array using the bubble sort algorithm.
 * 
 * @param {Array} arr - Array of comparable elements to be sorted
 * @param {Function} compareFn - Optional compare function for custom sorting
 * @returns {Array} Sorted array in ascending order
 * 
 * @example
 * bubbleSort([64, 34, 25, 12, 22, 11, 90])
 * // Returns: [11, 12, 22, 25, 34, 64, 90]
 * 
 * @example
 * bubbleSort(['banana', 'apple', 'cherry'])
 * // Returns: ['apple', 'banana', 'cherry']
 */
function bubbleSort(arr, compareFn = (a, b) => a > b) {
    if (!Array.isArray(arr) || arr.length <= 1) {
        return Array.isArray(arr) ? [...arr] : [];
    }
    
    // Create a copy to avoid modifying the original array
    const result = [...arr];
    const n = result.length;
    
    // Traverse through all array elements
    for (let i = 0; i < n; i++) {
        // Flag to optimize by detecting if array is already sorted
        let swapped = false;
        
        // Last i elements are already in place
        for (let j = 0; j < n - i - 1; j++) {
            // Compare and swap if elements are in wrong order
            if (compareFn(result[j], result[j + 1])) {
                [result[j], result[j + 1]] = [result[j + 1], result[j]];
                swapped = true;
            }
        }
        
        // If no swapping happened, array is sorted
        if (!swapped) {
            break;
        }
    }
    
    return result;
}

/**
 * Sorts an array in descending order using bubble sort.
 * 
 * @param {Array} arr - Array of comparable elements to be sorted
 * @returns {Array} Sorted array in descending order
 */
function bubbleSortDescending(arr) {
    return bubbleSort(arr, (a, b) => a < b);
}

/**
 * Bubble sort with step-by-step visualization.
 * 
 * @param {Array} arr - Array to be sorted
 * @returns {Object} Object containing sorted array and steps taken
 */
function bubbleSortVerbose(arr) {
    if (!Array.isArray(arr) || arr.length <= 1) {
        return {
            sortedArray: Array.isArray(arr) ? [...arr] : [],
            steps: 0
        };
    }
    
    const result = [...arr];
    const n = result.length;
    let steps = 0;
    
    console.log(`Initial array: [${result.join(', ')}]`);
    
    for (let i = 0; i < n; i++) {
        let swapped = false;
        console.log(`\nPass ${i + 1}:`);
        
        for (let j = 0; j < n - i - 1; j++) {
            steps++;
            if (result[j] > result[j + 1]) {
                [result[j], result[j + 1]] = [result[j + 1], result[j]];
                swapped = true;
                console.log(`  Swapped ${result[j + 1]} and ${result[j]}: [${result.join(', ')}]`);
            }
        }
        
        if (!swapped) {
            console.log('  No swaps needed. Array is sorted!');
            break;
        } else {
            console.log(`  End of pass ${i + 1}: [${result.join(', ')}]`);
        }
    }
    
    console.log(`\nFinal sorted array: [${result.join(', ')}]`);
    return {
        sortedArray: result,
        steps: steps
    };
}

/**
 * Async version of bubble sort with delays for visualization.
 * 
 * @param {Array} arr - Array to be sorted
 * @param {number} delay - Delay in milliseconds between each step
 * @param {Function} onStep - Callback function called after each step
 * @returns {Promise<Array>} Promise resolving to sorted array
 */
async function bubbleSortAsync(arr, delay = 100, onStep = null) {
    if (!Array.isArray(arr) || arr.length <= 1) {
        return Array.isArray(arr) ? [...arr] : [];
    }
    
    const result = [...arr];
    const n = result.length;
    
    for (let i = 0; i < n; i++) {
        let swapped = false;
        
        for (let j = 0; j < n - i - 1; j++) {
            if (result[j] > result[j + 1]) {
                [result[j], result[j + 1]] = [result[j + 1], result[j]];
                swapped = true;
                
                if (onStep) {
                    onStep([...result], i, j);
                }
                
                if (delay > 0) {
                    await new Promise(resolve => setTimeout(resolve, delay));
                }
            }
        }
        
        if (!swapped) {
            break;
        }
    }
    
    return result;
}

/**
 * Bubble sort for objects with a specified key.
 * 
 * @param {Array} arr - Array of objects to be sorted
 * @param {string} key - Key to sort by
 * @param {boolean} ascending - Sort in ascending order if true
 * @returns {Array} Sorted array of objects
 */
function bubbleSortBy(arr, key, ascending = true) {
    const compareFn = ascending 
        ? (a, b) => a[key] > b[key]
        : (a, b) => a[key] < b[key];
    
    return bubbleSort(arr, compareFn);
}

/**
 * Performance test for bubble sort.
 * 
 * @param {Array} arr - Array to sort
 * @returns {Object} Performance metrics
 */
function bubbleSortPerformance(arr) {
    const startTime = performance.now();
    const result = bubbleSort(arr);
    const endTime = performance.now();
    
    return {
        sortedArray: result,
        executionTime: endTime - startTime,
        arraySize: arr.length,
        timePerElement: (endTime - startTime) / arr.length
    };
}

// Export functions for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        bubbleSort,
        bubbleSortDescending,
        bubbleSortVerbose,
        bubbleSortAsync,
        bubbleSortBy,
        bubbleSortPerformance
    };
}

// Example usage and testing (only runs in browser or when file is run directly)
if (typeof window !== 'undefined' || (typeof require !== 'undefined' && require.main === module)) {
    console.log('=== Bubble Sort Examples ===');
    
    const testArrays = [
        [64, 34, 25, 12, 22, 11, 90],
        [5, 2, 8, 1, 9],
        [1, 2, 3, 4, 5],  // Already sorted
        [5, 4, 3, 2, 1],  // Reverse sorted
        [42],             // Single element
        [],               // Empty array
        [3, 3, 3, 3]      // All same elements
    ];
    
    testArrays.forEach((arr, index) => {
        console.log(`\nTest ${index + 1}: [${arr.join(', ')}]`);
        const sorted = bubbleSort(arr);
        console.log(`Sorted:  [${sorted.join(', ')}]`);
    });
    
    console.log('\n=== Verbose Example ===');
    bubbleSortVerbose([64, 34, 25, 12, 22, 11, 90]);
    
    console.log('\n=== Descending Order Example ===');
    const arr = [64, 34, 25, 12, 22, 11, 90];
    const descSorted = bubbleSortDescending(arr);
    console.log(`Original:   [${arr.join(', ')}]`);
    console.log(`Descending: [${descSorted.join(', ')}]`);
    
    console.log('\n=== String Sorting Example ===');
    const strings = ['banana', 'apple', 'cherry', 'date'];
    const sortedStrings = bubbleSort(strings);
    console.log(`Original: [${strings.join(', ')}]`);
    console.log(`Sorted:   [${sortedStrings.join(', ')}]`);
    
    console.log('\n=== Object Sorting Example ===');
    const people = [
        { name: 'Alice', age: 30 },
        { name: 'Bob', age: 25 },
        { name: 'Charlie', age: 35 }
    ];
    const sortedByAge = bubbleSortBy(people, 'age');
    console.log('Sorted by age:', sortedByAge);
    
    console.log('\n=== Performance Test ===');
    const perfArray = Array.from({ length: 1000 }, () => Math.floor(Math.random() * 1000));
    const perfResult = bubbleSortPerformance(perfArray);
    console.log(`Sorted ${perfResult.arraySize} elements in ${perfResult.executionTime.toFixed(2)}ms`);
    console.log(`Average time per element: ${perfResult.timePerElement.toFixed(4)}ms`);
}

File: algorithms/sorting/bubble-sort/implementations/c/bubble_sort.c
/**
 * Bubble Sort Implementation in C
 * 冒泡排序的C语言实现
 */

#include "bubble_sort.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/**
 * Swap two integers
 * 交换两个整数
 */
static void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

/**
 * Sort an integer array using bubble sort algorithm
 * 使用冒泡排序算法对整数数组进行排序
 */
bool bubble_sort(int* arr, size_t length) {
    if (arr == NULL || length == 0) {
        return true;  // Empty array is considered sorted
    }

    // Traverse through all array elements
    for (size_t i = 0; i < length; i++) {
        // Flag to optimize by detecting if array is already sorted
        bool swapped = false;

        // Last i elements are already in place
        for (size_t j = 0; j < length - i - 1; j++) {
            // Swap if the element found is greater than the next element
            if (arr[j] > arr[j + 1]) {
                swap(&arr[j], &arr[j + 1]);
                swapped = true;
            }
        }

        // If no swapping happened, array is sorted
        if (!swapped) {
            break;
        }
    }

    return true;
}

/**
 * Sort an integer array in descending order using bubble sort
 * 使用冒泡排序算法对整数数组进行降序排序
 */
bool bubble_sort_descending(int* arr, size_t length) {
    if (arr == NULL || length == 0) {
        return true;  // Empty array is considered sorted
    }

    for (size_t i = 0; i < length; i++) {
        bool swapped = false;

        for (size_t j = 0; j < length - i - 1; j++) {
            // Change comparison for descending order
            if (arr[j] < arr[j + 1]) {
                swap(&arr[j], &arr[j + 1]);
                swapped = true;
            }
        }

        if (!swapped) {
            break;
        }
    }

    return true;
}

/**
 * Sort an integer array with detailed result information
 * 对整数数组进行排序并返回详细信息
 */
bool bubble_sort_with_result(int* arr, size_t length, BubbleSortResult* result) {
    if (arr == NULL || result == NULL) {
        return false;
    }

    result->array = arr;
    result->length = length;
    result->comparisons = 0;
    result->swaps = 0;

    if (length == 0) {
        return true;  // Empty array is considered sorted
    }

    for (size_t i = 0; i < length; i++) {
        bool swapped = false;

        for (size_t j = 0; j < length - i - 1; j++) {
            result->comparisons++;
            if (arr[j] > arr[j + 1]) {
                swap(&arr[j], &arr[j + 1]);
                result->swaps++;
                swapped = true;
            }
        }

        if (!swapped) {
            break;
        }
    }

    return true;
}

/**
 * Generic bubble sort for any data type with custom comparison
 * 通用冒泡排序，支持自定义比较函数
 */
bool bubble_sort_generic(void* base, size_t num_elements, size_t element_size,
                        int (*compare)(const void*, const void*)) {
    if (base == NULL || num_elements == 0 || element_size == 0 || compare == NULL) {
        return true;  // Empty array is considered sorted
    }

    char* arr = (char*)base;

    for (size_t i = 0; i < num_elements; i++) {
        bool swapped = false;

        for (size_t j = 0; j < num_elements - i - 1; j++) {
            char* elem1 = arr + j * element_size;
            char* elem2 = arr + (j + 1) * element_size;

            if (compare(elem1, elem2) > 0) {
                // Swap elements
                for (size_t k = 0; k < element_size; k++) {
                    char temp = elem1[k];
                    elem1[k] = elem2[k];
                    elem2[k] = temp;
                }
                swapped = true;
            }
        }

        if (!swapped) {
            break;
        }
    }

    return true;
}

/**
 * Check if an array is sorted in ascending order
 * 检查数组是否按升序排序
 */
bool is_sorted_ascending(const int* arr, size_t length) {
    if (arr == NULL || length <= 1) {
        return true;
    }

    for (size_t i = 0; i < length - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }

    return true;
}

/**
 * Check if an array is sorted in descending order
 * 检查数组是否按降序排序
 */
bool is_sorted_descending(const int* arr, size_t length) {
    if (arr == NULL || length <= 1) {
        return true;
    }

    for (size_t i = 0; i < length - 1; i++) {
        if (arr[i] < arr[i + 1]) {
            return false;
        }
    }

    return true;
}

/**
 * Create a copy of an integer array
 * 创建整数数组的副本
 */
int* array_copy(const int* source, size_t length) {
    if (source == NULL || length == 0) {
        return NULL;
    }

    int* copy = (int*)malloc(length * sizeof(int));
    if (copy == NULL) {
        return NULL;
    }

    memcpy(copy, source, length * sizeof(int));
    return copy;
}

/**
 * Print an integer array
 * 打印整数数组
 */
void print_array(const int* arr, size_t length) {
    if (arr == NULL) {
        printf("NULL");
        return;
    }

    printf("[");
    for (size_t i = 0; i < length; i++) {
        if (i > 0) {
            printf(", ");
        }
        printf("%d", arr[i]);
    }
    printf("]");
}